backend/pkg/k8cache/k8cache.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	return cachedData, nil
+}
+
+// This function is used when serving response from cache to ensure the client
+// receives correct metadata about the response.
+func SetHeader(cacheData CachedResponseData, w http.ResponseWriter) {
+	for idx, header := range cacheData.Headers {
+		w.Header()[idx] = header
+	}
+
+	w.WriteHeader(cacheData.StatusCode)
+}
+
+// MarshallToStore serialize a cacheResponseData struct into JSON []byte.
+// This function is used before storing the K8's response data into cache.
+// ensuring a consistent and structured format for all cached entries.
+func MarshalToStore(cacheData CachedResponseData) ([]byte, error) {
+	jsonByte, err := json.Marshal(cacheData)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonByte, nil
+}
+
+const gzipEncoding = "gzip"
+
+// This ensures that the cached headers accurately reflect the state of the
+// decompressed body that is being stored, and prevents client side decompression
+// issues serving from cache.
+func SetHeadersToCache(responseHeaders http.Header, encoding string) http.Header {
+	cacheHeader := make(http.Header)
+
+	for idx, header := range responseHeaders {
+		if strings.EqualFold(idx, "Content-Encoding") && encoding == gzipEncoding {
+			continue
+		}
+
+		cacheHeader[idx] = append(cacheHeader[idx], header...)
+	}
+
+	return cacheHeader
+}
+
+var (
+	clientsetCache = make(map[string]*kubernetes.Clientset)
+	mu             sync.Mutex
+)
+
+// getClientMD is used to get a clientset for the given context and token.
+// It will reuse clientsets if a matching one is already cached.
+func getClientMD(k *kubeconfig.Context, token string) (*kubernetes.Clientset, error) {
+	contextKey := strings.Split(k.ClusterID, "+")
+	cacheKey := fmt.Sprintf("%s-%s", contextKey[1], hashToken(token))
+
+	mu.Lock()
+	defer mu.Unlock()
+
+	if cs, found := clientsetCache[cacheKey]; found {
+		return cs, nil
+	}
+
+	cs, err := k.ClientSetWithToken(token)
+	if err != nil {
+		return nil, fmt.Errorf("error while creating clientset for key %s: %w", cacheKey, err)
+	}
+
+	clientsetCache[cacheKey] = cs
+
+	return cs, nil
+}
+
+func hashToken(token string) string {
+	if len(token) > 32 {
+		return token[:32]
+	}
+
+	return token
+}
+
+// This function checks the user's permission to access the resource.
+// If the user is authorized and has permission to view the resources, it returns true.
+// Otherwise, it returns false if authorization fails.
+func IsAllowed(url *url.URL,
+	k *kubeconfig.Context,
+	w http.ResponseWriter,
+	r *http.Request,
+) (bool, error) {
+	token := r.Header.Get("Authorization")
+
+	clientset, err := getClientMD(k, token)
+	if err != nil {
+		return false, err
+	}
+
+	apiPath := mux.Vars(r)["api"]
+	parts := strings.Split(apiPath, "/")
+
+	last := parts[len(parts)-1]
+
+	review := &authorizationv1.SelfSubjectAccessReview{
+		Spec: authorizationv1.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationv1.ResourceAttributes{
+				Resource: last,
+				Verb:     "get",
```

Comment:
why is the verb hardcoded to `get` here? the request can also be for listing/watching the resource.

Reply:
Thanks